File: web/src/platform/capabilities.ts
export interface PlatformCapabilities {
  native: boolean;
  filePicker: boolean;
  folderPicker: boolean;
  notifications: boolean;
  tray: boolean;
  nativeMenus: boolean;
  deepLinks: boolean;
  autoStart: boolean;
  windowManagement: boolean;
}

export interface FilePickerOptions {
  title?: string;
  defaultPath?: string;
  filters?: Array<{
    name: string;
    extensions: string[];
  }>;
}

export interface NotificationOptions {
  title: string;
  body: string;
  icon?: string;
  silent?: boolean;
}

export interface DesktopBootstrapInfo {
  serverUrl: string;
  sessionToken: string;
  csrfToken: string;
}

File: internal/desktop/capabilities.go
package desktop

import (
	"os/exec"
	"runtime"
)

// Capabilities represents the native platform features supported by Nexus Desktop.
type Capabilities struct {
	Native           bool `json:"native"`
	FilePicker       bool `json:"filePicker"`
	FolderPicker     bool `json:"folderPicker"`
	Notifications    bool `json:"notifications"`
	Tray             bool `json:"tray"`
	NativeMenus      bool `json:"nativeMenus"`
	DeepLinks        bool `json:"deepLinks"`
	AutoStart        bool `json:"autoStart"`
	WindowManagement bool `json:"windowManagement"`
}

// DefaultCapabilities returns the standard native desktop capabilities.
func DefaultCapabilities() Capabilities {
	filePicker := false
	folderPicker := false
	notifications := false
	switch runtime.GOOS {
	case "darwin":
		filePicker = commandAvailable("osascript")
		folderPicker = filePicker
		notifications = filePicker
	case "linux":
		filePicker = commandAvailable("zenity")
		folderPicker = filePicker
		notifications = commandAvailable("notify-send")
	}

	return Capabilities{
		Native:           true,
		FilePicker:       filePicker,
		FolderPicker:     folderPicker,
		Notifications:    notifications,
		Tray:             false,
		NativeMenus:      false,
		DeepLinks:        false,
		AutoStart:        false,
		WindowManagement: true,
	}
}

func commandAvailable(name string) bool {
	_, err := exec.LookPath(name)
	return err == nil
}
